software/firmware/bootloader.py

```diff
@@ -144,4 +163,20 @@ def main(self):
             europi.b1._handler_both(europi.b2, self.exit_to_menu)
             europi.b2._handler_both(europi.b1, self.exit_to_menu)
 
-            script_class().main()
+            try:
+                script_class().main()
+            except Exception as err:
+                # set all outputs to zero for safety
+                europi.turn_off_all_cvs()
+
+                # in case we have the USB cable connected, print the stack trace for debugging
+                # otherwise, just halt and show the error message
+                print(f"[ERR ] Failed to run script: {err}")
+                sys.print_exception(err)
+
+                # show the type & first portion of the exception on the OLED
+                # we can only fit so many characters, so truncate as needed
+                MAX_CHARS = OLED_WIDTH // CHAR_WIDTH
+                self.show_error(
+                    "Crash", f"{str(type(err))[0:MAX_CHARS]}\n{str(err)[0:MAX_CHARS]}", -1
```

Comment:
[optional] `type(err)` returns a string like `<class 'Exception'>` which has a bunch of unnecessary characters. Consider using `type(err).__name__` instead, to print just the exception class' name. Or maybe `err.__class__.__name__` which reads slightly nicer, if you can ignore the underscores.

Reply:
I opted for version 2, with the underscores. I don't mind them, personally.